src/interface/viewport.c

```diff
@@ -844,6 +846,258 @@ void sub_0x69E8B0(uint32 eax, uint32 ecx){
+	{
+		RCT2_GLOBAL(0x141E9B4, uint32_t) = 0xFFFF;
+		RCT2_GLOBAL(0x141E9B8, uint32_t) = 0xFFFF;
+		RCT2_GLOBAL(0x141E9BC, uint32_t) = 0xFFFF;
+		RCT2_GLOBAL(0x141E9C0, uint32_t) = 0xFFFF;
+		RCT2_GLOBAL(0x141E9C4, uint32_t) = 0xFFFF;
+		RCT2_GLOBAL(0x141E9C8, uint32_t) = 0xFFFF;
+		RCT2_GLOBAL(0x141E9CC, uint32_t) = 0xFFFF;
+		RCT2_GLOBAL(0x141E9D0, uint32_t) = 0xFFFF;
+		RCT2_GLOBAL(0x141E9D4, uint32_t) = 0xFFFF;
+		RCT2_GLOBAL(0x141E9D8, uint32_t) = 0xFFFF;
+		RCT2_GLOBAL(0x141E9DC, uint32_t) = 0xFFFF;
+		//loc_68B3FB: Another function jumps to here. We need to split this!
+		RCT2_GLOBAL(0x141F56A, uint16_t) = 0;
+		RCT2_GLOBAL(0x9E3138, uint8_t) = 0xFF;
+		RCT2_GLOBAL(0x9E30B6, uint8_t) = 0xFF;
+		RCT2_GLOBAL(0x9E323C, uint8_t) = 0xFF;
+		RCT2_GLOBAL(0x9DE56A, uint16_t) = ax;
+		RCT2_GLOBAL(0x9DE56E, uint16_t) = cx;
+		RCT2_GLOBAL(0x9DE574, uint16_t) = ax;
+		RCT2_GLOBAL(0x9DE576, uint16_t) = cx;
+		int dx = cx;
+		int esi = dx;
+		esi <<= 8;
+		esi |= ax;
+		esi >>= 3;
+		int ax_tmp = ax;
+		int cx_tmp = cx;
+		rct_map_element* map_element = RCT2_ADDRESS(RCT2_ADDRESS_TILE_MAP_ELEMENT_POINTERS, rct_map_element*)[esi / 4];//8];
+		rct_drawpixelinfo* dpi = RCT2_GLOBAL(0x140E9A8, rct_drawpixelinfo*);
+		switch (RCT2_GLOBAL(RCT2_ADDRESS_CURRENT_ROTATION, uint32_t))
+		{
+		case 0:
+			dx = ax + cx;
+			break;
+		case 1:
+			ax += 32;
+			dx = cx - ax;
+			break;
+		case 2:
+			ax += 32;
+			cx += 32;
+			dx = -(ax + cx);
+			break;
+		case 3:
+			cx += 32;
+			dx = ax - cx;
+			break;
+		}
+		dx /= 2;
+		if ((RCT2_GLOBAL(RCT2_ADDRESS_MAP_SELECTION_FLAGS, uint16_t) & 4) &&
+			RCT2_GLOBAL(0x9DE56A, uint16_t) == RCT2_GLOBAL(0x9DEA48, uint16_t) &&
+			RCT2_GLOBAL(0x9DE56E, uint16_t) == RCT2_GLOBAL(0x9DEA4A, uint16_t))
+		{
+
+			int ebx = RCT2_GLOBAL(RCT2_ADDRESS_CURRENT_ROTATION, uint32_t);
+			RCT2_GLOBAL(0x9DE568, uint16_t) = ax;
+			RCT2_GLOBAL(0x9DE56C, uint16_t) = cx;
+			int dl = RCT2_GLOBAL(0x009DEA4E, uint8_t) & 3;
+			ebx += dl;
+			ebx &= 3;
+			dl = RCT2_GLOBAL(0x009DEA4E, uint8_t) & 0xFC;
+			ebx += dl;
+			ebx += 0x20900C27;
+			int d = RCT2_GLOBAL(0x009DEA4C, uint16_t);
+			RCT2_GLOBAL(0x9DE570, uint8_t) = 0;
+			RCT2_GLOBAL(0x9DEA52, uint16_t) = 0;
+			RCT2_GLOBAL(0x9DEA54, uint16_t) = 0;
+			RCT2_GLOBAL(0x9DEA56, uint16_t) = d + 18;
+			RCT2_CALLPROC_X(
+				(int)RCT2_ADDRESS(0x0098197C, uint32_t*)[RCT2_GLOBAL(RCT2_ADDRESS_CURRENT_ROTATION, uint32_t)],
+				0xFF00, ebx, cx & 0xFF00, d, 32, 32, RCT2_GLOBAL(RCT2_ADDRESS_CURRENT_ROTATION, uint32_t));
+		}
+		int bx = dx + 52;
+		if (bx > dpi->y)
+		{
+			rct_map_element* element = map_element;//push map_element
+			bx = element->clearance_height;
+			if (!(element->flags & MAP_ELEMENT_FLAG_LAST_TILE))
+			{
+				while (true)
+				{
+					element++;
+					if (bx < element->clearance_height) 
+						bx = element->clearance_height;
+					if (element->flags & MAP_ELEMENT_FLAG_LAST_TILE) break;
+				}
+			}
+			if ((element->type & 0x3C) == MAP_ELEMENT_TYPE_SURFACE &&
```

Comment:
Use the map element helper functions. There is one for getting the type and one for checking the last tile.